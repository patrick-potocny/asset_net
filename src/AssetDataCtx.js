import { createContext } from "react";

export const AssetDataCtx = createContext()