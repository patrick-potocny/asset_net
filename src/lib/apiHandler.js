import {
  processCryptoData,
  processCryptoSearch,
  cryptoTimeFrames,
  stocksTimeFrames,
  processStocksData,
  processStocksSearch,
  sortAssetData,
} from "./utlis";
import axios from "axios";
// import { createKeyIterator } from "./utlis";

// const keyIterator = createKeyIterator()

async function makeApiCall(options) {
  try {
    const response = await axios.request(options);
    return response;
  } catch (e) {
    if (e.response.status === 429) {
      alert(
        "Too many requests\nThis app uses free APIs that limits the requests per minute.\nTry again in a minute"
      );
    }
    return {};
  }
}

async function getAssetsData() {
  const assetList = JSON.parse(localStorage.getItem("assetList"));
  let assetData = [];

  await Promise.all(
    assetList.map(async (asset) => {
      const data = await getAssetData(asset);
      assetData.push(data);
    })
  );

  sortAssetData(assetData);
  return assetData;
}

async function getAssetData(asset) {
  let assetData = {};
  switch (asset.assetType) {
    case "crypto": {
      const options = {
        method: "GET",
        url: `https://coinranking1.p.rapidapi.com/coin/${asset.id}`,
        params: { timePeriod: cryptoTimeFrames[asset.timeFrame] },
        headers: {
          "X-RapidAPI-Key": process.env.REACT_APP_API_KEY,
          "X-RapidAPI-Host": "coinranking1.p.rapidapi.com",
        },
      };

      const data = await makeApiCall(options);
      assetData = processCryptoData(data.data, asset);
      break;
    }
    case "stocks": {
      const options = {
        method: "GET",
        url: "https://alpha-vantage.p.rapidapi.com/query",
        params: {
          function: stocksTimeFrames[asset.timeFrame],
          symbol: asset.id,
          datatype: "json",
          output_size: "compact",
        },
        headers: {
          "X-RapidAPI-Key": process.env.REACT_APP_API_KEY,
          "X-RapidAPI-Host": "alpha-vantage.p.rapidapi.com",
        },
      };

      if (options.params.function === "TIME_SERIES_INTRADAY")
        options.params.interval = "60min";

      const data = await makeApiCall(options);
      assetData = processStocksData(data.data, asset);
      break;
    }
  }
  return assetData;
}

async function  getSearchResults(query, assetType) {
  switch (assetType) {
    case "crypto": {
      const options = {
        method: "GET",
        url: "https://coinranking1.p.rapidapi.com/search-suggestions",
        params: { query: query },
        headers: {
          "X-RapidAPI-Key": process.env.REACT_APP_API_KEY,
          "X-RapidAPI-Host": "coinranking1.p.rapidapi.com",
        },
      };

      const response = await makeApiCall(options);
      return processCryptoSearch(response.data);
    }
    case "stocks": {
      const options = {
        method: "GET",
        url: "https://alpha-vantage.p.rapidapi.com/query",
        params: {
          keywords: query,
          function: "SYMBOL_SEARCH",
          datatype: "json",
        },
        headers: {
          "X-RapidAPI-Key": process.env.REACT_APP_API_KEY,
          "X-RapidAPI-Host": "alpha-vantage.p.rapidapi.com",
        },
      };

      const response = await makeApiCall(options);
      return processStocksSearch(response.data);
    }
  }
}

export { getAssetsData, getAssetData, getSearchResults };
