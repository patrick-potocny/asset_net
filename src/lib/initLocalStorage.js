export default function initLocalStorage() {
  // Check if local storage has been created 
  if (localStorage.getItem('darkMode') === null) {
    localStorage.setItem('darkMode', true)
  }
}