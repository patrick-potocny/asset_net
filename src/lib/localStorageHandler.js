function initLocalStorage() {
  // Check if local storage has been created
  if (localStorage.getItem("darkMode") === null) {
    localStorage.setItem("darkMode", false);
    localStorage.setItem("assetType", "all_assets");
    localStorage.setItem(
      "assetList",
      JSON.stringify([
        {
          id: "Qwsogvtv82FCd",
          timeFrame: "1D",
          assetType: "crypto",
        },
      ])
    );
  }
}

// Updates assets timeFrame
function updateAsset(id, newTf) {
  const assetList = JSON.parse(localStorage.getItem("assetList"));
  let updatedAsset = {};

  const updatedData = assetList.map((item) => {
    if (item.id === id) {
      updatedAsset = { ...item, timeFrame: newTf };
      return updatedAsset;
    }
    return item;
  });
  localStorage.setItem("assetList", JSON.stringify(updatedData));

  return updatedAsset;
}

function deleteAsset(id) {
  const assetList = JSON.parse(localStorage.getItem("assetList"));
  const updatedData = assetList.filter(item => item.id !== id);
  localStorage.setItem("assetList", JSON.stringify(updatedData));
}

function addAsset(id, name, assetType) {
  const assetList = JSON.parse(localStorage.getItem("assetList"));
  // Check if asset is already in localStorage
  if (assetList.filter(asset => asset.id === id).length > 0) return false

  const newAsset = {
    id: id,
    name: name,
    timeFrame: '1D',
    assetType: assetType
  }
  assetList.push(newAsset)
  localStorage.setItem("assetList", JSON.stringify(assetList));
  return newAsset
}

export {initLocalStorage, updateAsset, deleteAsset, addAsset};
