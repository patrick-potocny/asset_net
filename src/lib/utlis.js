const cryptoTimeFrames = {
  "1D": "24h",
  "1W": "7d",
  "1M": "30d",
  "3M": "3m",
  "1Y": "1y",
  "3Y": "3y",
};

function processCryptoData(data, asset) {
  let processedData = {
    symbol: data.data.coin.symbol,
    name: data.data.coin.name,
    price: parseFloat(data.data.coin.price).toFixed(2),
    change: data.data.coin.change,
    rising: true,
    id: asset.id,
    assetType: asset.assetType,
    timeFrame: asset.timeFrame,
    sparkline: {
      labels: generateCryptoLabels(asset.timeFrame),
      data: Object.values(data.data.coin.sparkline),
    },
  };

  if (processedData.change.includes("-")) {
    processedData.rising = false;
    processedData.change = processedData.change.substring(1);
  }

  return processedData;
}

function generateCryptoLabels(timeFrame) {
  let labels = [];
  const currentTime = new Date();

  switch (timeFrame) {
    case "1D":
      for (let i = 0; i < 24; i++) {
        const timestamp = new Date(currentTime);
        timestamp.setHours(timestamp.getHours() - 1 * i);
        timestamp.setMinutes(0);
        timestamp.setSeconds(0);
        labels.push(`${timestamp.getHours()}:00`);
      }
      break;

    case "1W":
      for (let i = 0; i < 7; i++) {
        const date = new Date();
        date.setDate(date.getDate() - i);
        const dayName = date.toLocaleString("default", { weekday: "short" });
        for (let i = 0; i < 3; i++) {
          labels.push(dayName.slice(0, 3));
        }
      }
      for (let i = 0; i < 3; i++) {
        labels.push(labels[labels.length - 1]);
      }
      break;

    case "1M":
      for (let i = 0; i < 24; i++) {
        const pastDate = new Date(
          currentTime.getTime() - i * 24 * 60 * 60 * 1000
        );
        const pastDateString = pastDate.toLocaleDateString("en-US", {
          month: "2-digit",
          day: "2-digit",
        });
        labels.push(pastDateString);
      }
      break;

    case "3M":
      for (let i = 0; i < 3; i++) {
        const date = new Date();
        date.setMonth(currentTime.getMonth() - i);
        const month = date.toLocaleString("default", { month: "long" });
        for (let i = 0; i < 8; i++) {
          labels.push(month.slice(0, 3));
        }
      }
      break;

    case "1Y":
      for (let i = 0; i < 12; i++) {
        const date = new Date();
        date.setMonth(currentTime.getMonth() - i);
        const month = date.toLocaleString("default", { month: "long" });
        for (let i = 0; i < 2; i++) {
          labels.push(month.slice(0, 3));
        }
      }
      break;

    case "3Y":
      for (let i = 0; i < 3; i++) {
        const date = new Date();
        date.setFullYear(currentTime.getFullYear() - i);
        const year = date.getFullYear();
        for (let i = 0; i < 8; i++) {
          labels.push(year);
        }
      }
      break;
  }

  labels.reverse().unshift("");
  return labels;
}

function processCryptoSearch(data) {
  const results = [];

  for (const coin of data.data.coins) {
    results.push({
      id: coin.uuid,
      symbol: coin.symbol,
      name: coin.name,
    });
  }

  return results.slice(0, 4);
}

function sortAssetData(assetData) {
  const sortBy = JSON.parse(localStorage.getItem("assetList"));
  assetData.sort(function (a, b) {
    return (
      sortBy.findIndex((item) => item.id === a.id) -
      sortBy.findIndex((item) => item.id === b.id)
    );
  });
}

export {
  processCryptoData,
  processCryptoSearch,
  cryptoTimeFrames,
  sortAssetData,
};
