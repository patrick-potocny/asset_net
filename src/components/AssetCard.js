import React, { useContext, useEffect, useState } from "react";
import PropTypes from "prop-types";
import LineChart from "./LineChart";
import Modal from "./Modal";
import DelAsset from "./DelAsset";
import removeIcon from "../assets/images/closeIcon.svg";
import greenTriangle from "../assets/images/greenTriangle.svg";
import redTriangle from "../assets/images/redTriangle.svg";
import { updateTf } from "../lib/utlis";
import { AssetDataCtx } from "../AssetDataCtx";

function AssetCard({ asset }) {
  const { assetData, setAssetData } = useContext(AssetDataCtx);
  const [triangle, setTriangle] = useState(greenTriangle);
  const [changeColor, setChangeColor] = useState("green");
  const [isOpen, setIsOpen] = useState(false);
  const timeFrames = ["1D", "1W", "1M", "3M", "1Y", "3Y"];

  useEffect(() => {
    // Check if change is negative and if so change color and arrow
    if (!asset.rising) {
      setTriangle(redTriangle);
      setChangeColor("red");
    }
  }, []);

  return (
    <div className="asset">
      <img
        className="asset__remove"
        onClick={() => setIsOpen(true)}
        data-id={asset.id}
        src={removeIcon}
        alt="remove"
      />
      <Modal isOpen={isOpen} onClose={() => setIsOpen(false)}>
        <DelAsset
          assetId={asset.id}
          assetSymbol={asset.symbol}
          onClose={() => setIsOpen(false)}
        />
      </Modal>
      <div className="asset__info">
        <p className="symbol">{asset.symbol}</p>
        <span className="name">{asset.name}</span>
        <hr className="info-divider" />
        <div className="flex">
          <p className="price">{asset.price} $</p>
          <div className="change">
            <img src={triangle} alt="triangle" className="change__triangle" />
            <p className={"change__number " + changeColor}>{asset.change}%</p>
          </div>
        </div>
      </div>
      <hr className="asset__divider" />
      <div className="asset__time-frame">
        {timeFrames.map((tf) => (
          <button
            key={tf}
            onClick={() => {
              updateTf(asset.id, tf, assetData, setAssetData);
            }}
            className={`time-frame-btn ${
              tf === asset.timeFrame ? "selected" : ""
            }`}
          >
            {tf}
          </button>
        ))}
      </div>
      <hr className="asset__divider" />
      <LineChart data={asset.sparkline} changeColor={changeColor} />
    </div>
  );
}

AssetCard.propTypes = {
  asset: PropTypes.object.isRequired,
};

export default AssetCard;
