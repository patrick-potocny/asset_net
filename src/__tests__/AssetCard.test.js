/* eslint-disable */
import React from "react";
import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import AssetCard from "../components/AssetCard";
import { AssetDataCtx } from "../AssetDataCtx";
import userEvent from "@testing-library/user-event";
import { updateTf } from "../lib/utlis";

jest.mock("../components/LineChart", () =>
  jest.fn(() => <div>Line Chart</div>)
);
jest.mock("../components/Modal", () => ({ isOpen, onClose }) => (
  <div data-testid="mock-modal">isOpen is {isOpen.toString()}</div>
));
jest.mock("../lib/utlis", () => {
  return {
    updateTf: jest.fn(),
  };
});

const asset = {
  symbol: "BTC",
  name: "Bitcoin",
  price: "21088.66",
  change: "1.15",
  rising: true,
  id: "Qwsogvtv82FCd",
  assetType: "crypto",
  timeFrame: "1D",
  sparkline: {
    labels: [],
    data: [],
  },
};

describe("AssetCard component", () => {
  test("should render correctly", () => {
    const { container } = render(
      <AssetDataCtx.Provider value={{ assetData: [], setAssetData: jest.fn() }}>
        <AssetCard asset={asset} />
      </AssetDataCtx.Provider>
    );
    expect(container).toMatchSnapshot();
  });

  test("should render modal when remove icon is clicked", () => {
    render(
      <AssetDataCtx.Provider value={{ assetData: [], setAssetData: jest.fn() }}>
        <AssetCard asset={asset} />
      </AssetDataCtx.Provider>
    );
    expect(screen.getByText("isOpen is false")).toBeInTheDocument();

    userEvent.click(screen.getByAltText("remove"));

    expect(screen.getByText("isOpen is true")).toBeInTheDocument();
  });

  test("should call updateTf function when button is clicked", () => {
    render(
      <AssetDataCtx.Provider value={{ assetData: [], setAssetData: jest.fn() }}>
        <AssetCard asset={asset} />
      </AssetDataCtx.Provider>
    );

    userEvent.click(screen.getByText("1M"));

    expect(updateTf).toHaveBeenCalled();
  });
});
