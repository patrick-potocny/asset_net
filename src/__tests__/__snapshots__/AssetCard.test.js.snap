// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AssetCard component should render correctly 1`] = `
<div>
  <div
    class="asset"
  >
    <img
      alt="remove"
      class="asset__remove"
      data-id="Qwsogvtv82FCd"
      src="closeIcon.svg"
    />
    <div
      data-testid="mock-modal"
    >
      isOpen is 
      false
    </div>
    <div
      class="asset__info"
    >
      <p
        class="symbol"
      >
        BTC
      </p>
      <span
        class="name"
      >
        Bitcoin
      </span>
      <hr
        class="info-divider"
      />
      <div
        class="flex"
      >
        <p
          class="price"
        >
          21088.66
           $
        </p>
        <div
          class="change"
        >
          <img
            alt="triangle"
            class="change__triangle"
            src="greenTriangle.svg"
          />
          <p
            class="change__number green"
          >
            1.15
            %
          </p>
        </div>
      </div>
    </div>
    <hr
      class="asset__divider"
    />
    <div
      class="asset__time-frame"
    >
      <button
        class="time-frame-btn selected"
      >
        1D
      </button>
      <button
        class="time-frame-btn "
      >
        1W
      </button>
      <button
        class="time-frame-btn "
      >
        1M
      </button>
      <button
        class="time-frame-btn "
      >
        3M
      </button>
      <button
        class="time-frame-btn "
      >
        1Y
      </button>
      <button
        class="time-frame-btn "
      >
        3Y
      </button>
    </div>
    <hr
      class="asset__divider"
    />
    <div>
      Line Chart
    </div>
  </div>
</div>
`;
